### PR TITLE
Support auto-login with JWT #162

### PR DESCRIPTION
**Pull request by anatol-sialitski**

### CODE REVIEW

**Review by rymsha — CHANGES_REQUESTED** [review]

File: build.gradle [hunk 1]
```diff
@@ -23,6 +23,10 @@ dependencies {
     include "com.enonic.lib:lib-http-client:3.2.2"
     include "com.nimbusds:oauth2-oidc-sdk:11.9.1"
 
+
+    include 'com.auth0:java-jwt:4.4.0'
```
[the remark this answers is not in the document]
Comment (anatol-sialitski) [reply build.gradle:27]: it will be done in the separate issue

File: docs/index.adoc [hunk 2]
```diff
@@ -182,12 +185,19 @@ idprovider.<idprovidername>.endSession.additionalParameters.0.value=  (string, r
...
 idprovider.<idprovidername>.defaultGroups=  (space separated group keys, optional)
 
 idprovider.<idprovidername>.rules.forceEmailVerification=  (true | false, required)
+
+idprovider.<idprovidername>.autoLogin.enforce= (true|false, optional)
+idprovider.<idprovidername>.autoLogin.createUsers=  (true|false, optional)
+idprovider.<idprovidername>.autoLogin.createSession=  (true|false, optional)
+# Look for a token in the Sec-WebSocket-Protocol header.
+idprovider.<idprovidername>.autoLogin.retrievalWsHeader=  (true|false, optional)
+idprovider.<idprovidername>.autoLogin.validationAllowedAudience=  (space separated strings, optional)
```
Comment (rymsha) [inline docs/index.adoc:200]:
```suggestion
idprovider.<idprovidername>.autoLogin.allowedAudience=  (space separated strings, optional)
```

File: docs/index.adoc [hunk 3]
```diff
@@ -182,12 +185,19 @@ idprovider.<idprovidername>.endSession.additionalParameters.0.value=  (string, r
...
 
 idprovider.<idprovidername>.defaultGroups=  (space separated group keys, optional)
 
 idprovider.<idprovidername>.rules.forceEmailVerification=  (true | false, required)
+
+idprovider.<idprovidername>.autoLogin.enforce= (true|false, optional)
+idprovider.<idprovidername>.autoLogin.createUsers=  (true|false, optional)
+idprovider.<idprovidername>.autoLogin.createSession=  (true|false, optional)
+# Look for a token in the Sec-WebSocket-Protocol header.
+idprovider.<idprovidername>.autoLogin.retrievalWsHeader=  (true|false, optional)
```
Comment (rymsha) [inline docs/index.adoc:199]:
```suggestion
idprovider.<idprovidername>.autoLogin.wsHeader=  (true|false, optional)
```

File: src/main/java/com/enonic/app/oidcidprovider/handler/IdProviderManager.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+import com.enonic.app.oidcidprovider.jwt.RSAAlgorithmProvider;
+
+public class IdProviderManager
+{
+    private static final int TIMEOUT_MS = 5000;
+
+    private final Map<String, Object> idProviderConfig;
+
+    private RSAAlgorithmProvider algorithmProvider;
```
Comment (rymsha) [inline src/main/java/com/enonic/app/oidcidprovider/handler/IdProviderManager.java:18]: make sure it is `final`.

File: src/main/java/com/enonic/app/oidcidprovider/handler/IdProviderConfigService.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+
+import com.enonic.app.oidcidprovider.jwt.RSAAlgorithmProvider;
+
+@Component(immediate = true, service = IdProviderConfigService.class, configurationPid = "com.enonic.app.oidcidprovider")
+public class IdProviderConfigService
+{
+    private static final Cache<String, IdProviderManager> CACHE = CacheBuilder.newBuilder().build();
```
Comment (rymsha) [inline src/main/java/com/enonic/app/oidcidprovider/handler/IdProviderConfigService.java:16]: ConcurrentHashMap is sufficient

File: src/main/resources/lib/login.js [hunk 6]
```diff
@@ -4,67 +4,38 @@ const configLib = require('/lib/config');
...
-    const loginResult = authLib.login({
-        user: userName,
-        idProvider: idProviderKey,
-        skipAuth: true
-    });
-    if (loginResult.authenticated) {
-        log.debug(`Logged in user [${principalKey}]`);
-    } else {
-        throw `Error while logging user [${principalKey}]`;
+        modifyUserProfile(claims, principalKey);
```
Comment (rymsha) [inline src/main/resources/lib/login.js:30]: do we always try to modify a user on every login? This should not be done at least for autologin

**Review by rymsha — CHANGES_REQUESTED** [review]

File: src/main/resources/lib/login.js (hunk 6 above)
Comment (anatol-sialitski) [reply src/main/resources/lib/login.js:30]: Are you sure? `modifyUserProfile` removes non supported claims.

**Review by rymsha — APPROVED** [review]

**Pull request merged**